15-task-object

siiiic opened the pull request.

siiiic committed:
15-task-object

GitEsense commented on 15-task-object/task-object.js:
```diff
@@ -0,0 +1,39 @@
+    getTask: function (id) {
+        return this.list.findIndex(task => task.id === id);
+    },
+    addTask: function (title, priority) {
+        const id = this.list.length + 1;
+        this.list.push({title, id, priority});
+    },
```
Лучше id присваивать не по длине массива, а по внешнему каунтеру (например свойство того же объекта lastId
```js
const toDoList = {
   tasks: [],
   lastId: 0,
   addTask: function (title, priority) {
       this.tasks.push({ title, id: ++this.lastId, priority });
  },
```
Так же можно провести проверки и в идеале вообще принимать на вход "объект" (там будет удобнее работать с множеством свойств)

GitEsense commented on 15-task-object/task-object.js:
```diff
@@ -0,0 +1,39 @@
+
+const toDoList = {
+    list: [],
+    getTask: function (id) {
+        return this.list.findIndex(task => task.id === id);
+    },
```
в данной функции можно реализовать поиск не индекса, а самой задачи и если она есть - возвращать её ссылку
```js
    getTaskById(taskId) {
        const task = this.tasks.find(({ id }) => id === taskId);
        if (!task) {
            console.log(`Задача с id ${taskId} еще не добавлена в ваш список дел.`);
        }
        return task;
    },
```

GitEsense commented on 15-task-object/task-object.js:
```diff
@@ -0,0 +1,39 @@
+        const id = this.list.length + 1;
+        this.list.push({title, id, priority});
+    },
+    deleteTask: function (id) {
+        this.getTask(id) === -1 ?
+        console.log(`Задачи с id: ${id} нет в списке`) :
+        this.list.splice(this.getTask(id), 1);
+    },
```
Если изменить метод getTask и получать задачу (да и в этой реализации так же), для более удобного чтения кода, можно реализовать "от отрицательного ожидания":
```js
 deleteTask: function (id) {
        const taskIndex = this.getTask(id);
        if(taskIndex === -1){
            console.log(`Задачи с id: ${id} нет в списке`);
            return;
        }
        this.list.splice(taskIndex, 1);
    },
```
Если реализация через getTaskById, то
```js
    deleteTask: function (id) {
        const task = this.getTaskById(id);
        if (!task) {
            return;
        }
        this.list = this.list.filter((f) => f.id === task.id);
    },
```
Но даже при Вашей реализации - лучше воспользоваться фильтром массива с присвоением, а не splice, т.к. splice будет более дорогостоящей операцией

GitEsense commented on 15-task-object/task-object.js:
```diff
@@ -0,0 +1,39 @@
+    sortByPriority: function () {
+        this.list.sort((a, b) => a.priority - b.priority);  
+    },
+    updateTitle: function (id, title) {
+        this.list[this.getTask(id)].title = title;
+    },
```
Лучше разбить на несколько строк, зато будет удобнее читать
```js
    updateTitle: function (id, title) {
        const task = this.getTaskById(id);
        if (!task) {
            return;
        }
        if(!title){
            console.log('Не могу заменить на пустое значение');
            return;
        }
        task.title = title;
        return this;
    },
```

GitEsense reviewed (approved):
В целом работает, но не очень "читабельный" код и отсутствуют проверки. Оставил рекомендации и примеры "доработки", так же можете (так будет лучше) - реализовать через передачу не title + priority, а объекта, изменённый формат уже сдавать в следующем ДЗ (если конечно захотите менять)

siiiic committed:
Правки

1. Изменил функицю getTask
2. ID присваивается по внешнему каунтеру
3. Добавил проверки
4. Объеденил функции изменения названия задачи и приоритета в одну
5. Сортировка по приоритету теперь работает в обе стороны
6. Добавил вывод сообщений
